### PR TITLE
Fix race condition bug during local session repair

### PR DESCRIPTION
**Pull request by bbotella**
Syncronizes session onSuccess handlePrepareSession for LocalSessions
patch by Bernardo Botella Corbi

### CODE REVIEW

File: src/java/org/apache/cassandra/repair/consistent/LocalSessions.java [hunk 1]
```diff
@@ -828,16 +828,19 @@ public void onSuccess(@Nullable List<Void> result)
             {
                 try
                 {
-                    logger.info("Prepare phase for incremental repair session {} completed", sessionID);
-                    if (session.getState() != FAILED)
-                        setStateAndSave(session, PREPARED);
-                    else
-                        logger.info("Session {} failed before anticompaction completed", sessionID);
+                    synchronized (session)
```
Comment (yifan-c) [inline src/java/org/apache/cassandra/repair/consistent/LocalSessions.java]: The synchronized scope is a bit too large. I think you only need to make the two operations, `getState` and `setStateAndSave` atomic. You can make the synchronized block to cover the if..else.. statement.

One small step forward, you can probably add a `prepareSessionExceptFailed` method to wrap the get and set logic, e.g.
```java
    /**
     * return true if prepared. Return false if not, i.e. session faield
     */
    boolean prepareSessionExceptFailed(LocalSession session)
```

Comment (bbotella) [reply src/java/org/apache/cassandra/repair/consistent/LocalSessions.java]: Both suggestions make sense. Posting a new commit to address them.

File: src/java/org/apache/cassandra/repair/consistent/LocalSessions.java [hunk 2]
```diff
@@ -829,14 +829,11 @@ public void onSuccess(@Nullable List<Void> result)
...
-                        setStateAndSave(session, PREPARED);
-                    else
+                    if (!prepareSessionExceptFailed(session))
                         logger.info("Session {} failed before anticompaction completed", sessionID);
-
                     Message<PrepareConsistentResponse> message =
-                        Message.out(PREPARE_CONSISTENT_RSP,
-                                    new PrepareConsistentResponse(sessionID, getBroadcastAddressAndPort(), session.getState() != FAILED));
+                    Message.out(PREPARE_CONSISTENT_RSP,
+                                new PrepareConsistentResponse(sessionID, getBroadcastAddressAndPort(), session.getState() != FAILED));
```
Comment (yifan-c) [inline src/java/org/apache/cassandra/repair/consistent/LocalSessions.java]: Can you restore the removed spaces? The change is not required for the fix.

File: CHANGES.txt [hunk 3]
```diff
@@ -1,3 +1,5 @@
+Flaky testNoSuchRepairSessionAnticompaction (CASSANDRA-17335)
```
Comment (yifan-c) [inline CHANGES.txt]: We do not normally add test-only fix to the changes list. The change in this patch is to fix a bug/race condition during repair. Can you rephrase the description? So it is more clear than "Flaky testNoSuchRepairSessionAnticompaction"

**Review by yifan-c — COMMENTED** [review]
a few more nits. Looks good overall

File: src/java/org/apache/cassandra/repair/consistent/LocalSessions.java (hunk 2 above)
Comment (bbotella) [reply src/java/org/apache/cassandra/repair/consistent/LocalSessions.java]: Yes. For some reason my IDE decided it was a good idea removing them. Sorry for this.

File: CHANGES.txt (hunk 3 above)
Comment (bbotella) [reply CHANGES.txt]: Done
